src/Optimizer/Transformer/OptimizeHeroImages.php

```diff
@@ -0,0 +1,709 @@
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if ($this->configuration->get(OptimizeHeroImagesConfiguration::OPTIMIZE_HERO_IMAGES) === false) {
+            return;
+        }
+
+        $maxHeroImageCount = $this->configuration->get(OptimizeHeroImagesConfiguration::MAX_HERO_IMAGE_COUNT);
+
+        $this->inlineStyleBackupAttribute = $this->configuration->get(
+            OptimizeHeroImagesConfiguration::INLINE_STYLE_BACKUP_ATTRIBUTE
+        );
+
+        $heroImages     = $this->findHeroImages($document, $maxHeroImageCount);
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > $maxHeroImageCount) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = $maxHeroImageCount;
+        }
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->removeLazyLoading($heroImages[$index]);
+            $this->generatePreload($heroImages[$index], $document, $errors);
+            $this->generateImg($heroImages[$index], $document);
+        }
+    }
+
+    /**
+     * Find the hero images to optimize.
+     *
+     * @param Document $document          Document to look for hero images in.
+     * @param int      $maxHeroImageCount Maximum number of hero images to accept.
+     * @return HeroImage[] Array of hero images to optimize.
+     */
+    private function findHeroImages(Document $document, $maxHeroImageCount)
+    {
+        $heroImages          = [];
+        $heroImageCandidates = [];
+        $heroImageFallback   = null;
+        $seenParagraphCount  = 0;
+        $node                = $document->body;
+
+        while ($node !== null) {
+            if (! $node instanceof Element) {
+                $node = NodeWalker::nextNode($node);
+                continue;
+            }
+
+            if ($node->tagName === Tag::P) {
+                $seenParagraphCount++;
+            }
+
+            $heroImage = $this->detectImageWithAttribute($node, Attribute::DATA_HERO);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            } elseif (count($heroImageCandidates) < $maxHeroImageCount) {
+                $heroImageCandidate = $this->detectImageWithAttribute($node, Attribute::DATA_HERO_CANDIDATE);
+                if ($heroImageCandidate) {
+                    $heroImageCandidates[] = $heroImageCandidate;
+                } elseif ($seenParagraphCount < 2 && ! $heroImageFallback) {
+                    $heroImageFallback = $this->detectPossibleHeroImageFallbacks($node);
```

Comment:
I simplified the logic in d85938a868f04ac8c921137951440db4d6f3218a since I found that we were collecting all the hero image fallbacks up until the maximum, but then we were only ever using the first image fallback as a hero.